component/prometheus/operator/podmonitors/crdmanager.go

```diff
@@ -0,0 +1,280 @@
+	if err != nil {
+		return fmt.Errorf("building label selector: %w", err)
+	}
+	for _, ns := range c.Args.Namespaces {
+		opts := cache.Options{
+			Scheme:    scheme,
+			Namespace: ns,
+		}
+
+		if ls != labels.Nothing() {
+			opts.DefaultSelector.Label = ls
+		}
+		cache, err := cache.New(config, opts)
+		if err != nil {
+			return err
+		}
+
+		informers := cache
+
+		go func() {
+			err := informers.Start(ctx)
+			if err != nil && ctx.Err() != nil {
+				level.Error(c.Logger).Log("msg", "failed to start informers", "err", err)
+			}
+		}()
+		if !informers.WaitForCacheSync(ctx) {
+			return fmt.Errorf("informer caches failed to sync")
+		}
+		if err := c.configureInformers(ctx, informers); err != nil {
+			return fmt.Errorf("failed to configure informers: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// configureInformers configures the informers for the CRDManager to watch for PodMonitors changes.
+func (c *CRDManager) configureInformers(ctx context.Context, informers cache.Informers) error {
+	objects := []client.Object{
+		&v1.PodMonitor{},
+	}
+
+	informerCtx, cancel := context.WithTimeout(ctx, informerSyncTimeout)
+	defer cancel()
+
+	for _, obj := range objects {
+		informer, err := informers.GetInformer(informerCtx, obj)
+		if err != nil {
+			if errors.Is(informerCtx.Err(), context.DeadlineExceeded) { // Check the context to prevent GetInformer returning a fake timeout
+				return fmt.Errorf("timeout exceeded while configuring informers. Check the connection"+
+					" to the Kubernetes API is stable and that the Agent has appropriate RBAC permissions for %v", obj)
+			}
+
+			return err
+		}
+		_, err = informer.AddEventHandler((toolscache.ResourceEventHandlerFuncs{
+			AddFunc:    c.onAddPodMonitor,
+			UpdateFunc: c.onUpdatePodMonitor,
+			DeleteFunc: c.onDeletePodMonitor,
+		}))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// apply applies the current state of the CRDManager to the Prometheus discovery manager and scrape manager.
+func (c *CRDManager) apply() error {
+	c.Mut.Lock()
+	defer c.Mut.Unlock()
+	err := c.DiscoveryManager.ApplyConfig(c.DiscoveryConfigs)
+	if err != nil {
+		level.Error(c.Logger).Log("msg", "error applying discovery configs", "err", err)
+		return err
+	}
+	scs := []*config.ScrapeConfig{}
+	for _, sc := range c.ScrapeConfigs {
+		scs = append(scs, sc)
+	}
+	err = c.ScrapeManager.ApplyConfig(&config.Config{
+		ScrapeConfigs: scs,
+	})
+	if err != nil {
+		level.Error(c.Logger).Log("msg", "error applying scrape configs", "err", err)
+		return err
+	}
+	level.Debug(c.Logger).Log("msg", "scrape config was updated")
+	return nil
+}
+
+func (c *CRDManager) clearConfigs(ns string, name string) {
+	c.Mut.Lock()
+	defer c.Mut.Unlock()
+	prefix := fmt.Sprintf("podMonitor/%s/%s", ns, name)
+	for k := range c.DiscoveryConfigs {
+		if strings.HasPrefix(k, prefix) {
+			delete(c.DiscoveryConfigs, k)
+			delete(c.ScrapeConfigs, k)
+		}
+	}
+	delete(c.DebugInfo, prefix)
+}
```

Comment:
It's be nice to move this function to be defined _after_ the onXPodMonitor methods, since I wasn't clear what it did at first.

Typically, I recommend sorting functions in terms of usage order (i.e., if function A calls function B, then define function A first). This makes code much easier to read top-down, as you don't have to hop around the file.